### PR TITLE
Implement singleElement for iterables

### LINKED ISSUE #1927: singleElement for iterables
#### Summary
Add a method ```hasSingleElementThat``` to the assertions API for iterables that checks whether the iterable ```hasSize(1)``` and then returns ```element(0)``` for further assertions on that element.

I have stumbled across this pattern several times, so I wonder whether such a convenience method would make sense in the API. 

#### Example

```java
assertThat(newArrayList("Yoda")).hasSingleElementThat().startsWith("Yo");
```

### PR DESCRIPTION
**Pull request by mgrafl**
#### Check List:
* Fixes #1927 
* Unit tests : YES
* Javadoc with a code example (on API only) : YES

I would appreciate feedback on the following topics: 
* The implementations for ```singleElement()``` and ```singleElement(InstanceOfAssertFactory)``` simply call the API methods for ```hasSize``` and ```first```. Is this okay or should we use the internal methods (```iterables.assertHasSize``` and ```internalFirst```) instead?
* I have noticed that various places (JavaDoc, unit tests) use different utilities for creating lists (e.g., ```newArrayList```, ```list```, ```asList```). Which utility method is preferred?
* Should ```hasOnlyOneElementSatisfying``` be deprecated?

### CODE REVIEW

**Comment by joel-costigliola** [comment]
Thanks @mgrafl, I will review your PR very shortly !

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java [hunk 1]
```diff
@@ -2804,6 +2804,105 @@ private ELEMENT_ASSERT internalElement(int index) {
...
+  /**
+   * Verifies that the {@link Iterable} under test contains a single element and allow to perform assertions that element.
+   * <p>
+   * This is a shorthand for <code>hasSize(1).first()</code>.
+   * <p>
+   * By default available assertions after {@code singleElement()} are {@code Object} assertions, it is possible though to
+   * get more specific assertions if you create {@code IterableAssert} with either:
+   * <ul>
+   * <li>the element assert class, see: {@link Assertions#assertThat(Iterable, Class) assertThat(Iterable, element assert class)}</li>
+   * <li>an assert factory used that knows how to create elements assertion, see: {@link Assertions#assertThat(Iterable, AssertFactory) assertThat(Iterable, element assert factory)}</li>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIterableAssert.java:2816]: Add `singleElement(InstanceOfAssertFactory<?, ASSERT> assertFactory)` as another available option here with a link to the method

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java [hunk 2]
```diff
@@ -2804,6 +2804,105 @@ private ELEMENT_ASSERT internalElement(int index) {
...
+   * </ul>
+   * <p>
+   * Example: default {@code Object} assertions
+   * <pre><code class='java'> List&lt;String&gt; babySimpsons = asList("Maggie");
+   *
+   * // assertion succeeds, only Object assertions are available after singleElement()
+   * assertThat(babySimpsons).singleElement()
+   *                    .isEqualTo("Maggie");
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIterableAssert.java:2824]: minor: align `.isEqualTo("Maggie");` with `.singleElement()`

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java [hunk 3]
```diff
@@ -2804,6 +2804,105 @@ private ELEMENT_ASSERT internalElement(int index) {
...
+   * <p>
+   * Example: use of {@code String} assertions after {@code singleElement(as(InstanceOfAssertFactories.STRING)}
+   * <pre><code class='java'> List&lt;String&gt; babySimpsons = asList("Maggie");
+   *
+   * // assertion succeeds
+   * assertThat(babySimpsons).singleElement(as(InstanceOfAssertFactories.STRING))
+   *                    .startsWith("Mag");
+   * 
+   * // assertion fails
+   * assertThat(babySimpsons).singleElement(as(InstanceOfAssertFactories.STRING))
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIterableAssert.java:2879]: on this one only, use `STRING` only just to show that with importing  statically `InstanceOfAssertFactories.STRING` makes the code to read nicely

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java (hunk 2 above)
Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractIterableAssert.java:2824]: that's a general comment on the various code examples

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java (hunk 1 above)
Comment (mgrafl) [reply src/main/java/org/assertj/core/api/AbstractIterableAssert.java:2818]: Thanks for the hint. It has been added.

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java (hunk 2 above)
Comment (mgrafl) [reply src/main/java/org/assertj/core/api/AbstractIterableAssert.java]: Thanks for spotting it.

**Comment by mgrafl** [comment]
Thanks for the review and the clarifications. 
I have included the minor remarks in the application code as well as the deprecation of ```hasOnlyOneElementSatisfying```. 
I will add the missing unit tests for soft assertions in the next days.

**Comment by mgrafl** [comment]
The suggested unit tests have been added. 
Since I have limited experience with soft assertions in assertj, I would appreciate a review whether the tests check the correct things, @joel-costigliola. 

I introduced separate tests for ```iterable_soft_assertions_should_work_with_singleElement_navigation``` and ```list_soft_assertions_should_work_with_singleElement_navigation``` (in ```SoftAssertionsTests``` and ```BDDSoftAssertionsTest```) since the navigation part of ```singleElement``` presumes a list with only one element and I wanted to keep the tests concise.

**Comment by joel-costigliola** [comment]
thanks @mgrafl, I'll look at your PR this week end.

**Comment by joel-costigliola** [comment]
Good work thanks @mgrafl!

**Pull request closed without merging**